File: scala/sdktest/OneTimePasswordEntityTest.scala
// Generated basic-flow test for the one_time_password entity (model-driven;
// mirrors the java TestEntity generator). A dependency-free scala-cli test
// object driven by SdkEntityTestMain. Runs against the in-memory test
// transport seeded with the shipped OneTimePasswordTestData.json fixtures.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}

import voxgig.thesmsworkssdk.core.{Helpers, ThesmsworksSDK}
import voxgig.thesmsworkssdk.utility.struct.Struct

object OneTimePasswordEntityTest {

  def run(rep: SdkTestReport): Unit = {
    rep.scope("one_time_password.instance") {
      val testsdk = ThesmsworksSDK.testSDK()
      val ent = testsdk.oneTimePassword(null)
      rep.check("one_time_password.instance", ent != null, "expected non-null one_time_password entity")
    }

    rep.scope("one_time_password.basic") {
      val entityData = Helpers.toMapAny(SdkTestSupport.readJson(
          "../.sdk/test/entity/one_time_password/OneTimePasswordTestData.json"))
      val options = new LinkedHashMap[String, Object]()
      options.put("entity", entityData.get("existing"))
      val client = ThesmsworksSDK.testSDK(options, null)

      val idmap = new LinkedHashMap[String, Object]()
      idmap.put("one_time_password01", "ONE_TIME_PASSWORD01")
      idmap.put("one_time_password02", "ONE_TIME_PASSWORD02")
      idmap.put("one_time_password03", "ONE_TIME_PASSWORD03")
      idmap.put("otp01", "OTP01")
      idmap.put("otp02", "OTP02")
      idmap.put("otp03", "OTP03")
      val now = System.currentTimeMillis()

      // CREATE
      val oneTimePasswordRef01Ent = client.oneTimePassword(null)
      var oneTimePasswordRef01Data = Helpers.toMapAny(Struct.getprop(
          Struct.getpath(entityData, "new.one_time_password"), "one_time_password_ref01"))
      val oneTimePasswordRef01DataResult = oneTimePasswordRef01Ent.create(oneTimePasswordRef01Data, null)
      oneTimePasswordRef01Data = Helpers.toMapAny(oneTimePasswordRef01DataResult)
      rep.check("one_time_password.create.map", oneTimePasswordRef01Data != null, "expected create result to be a map")

      // LOAD
      val oneTimePasswordRef01MatchDt0 = new LinkedHashMap[String, Object]()
      val oneTimePasswordRef01DataDt0Loaded = oneTimePasswordRef01Ent.load(oneTimePasswordRef01MatchDt0, null)
      rep.check("one_time_password.load.nonnull", oneTimePasswordRef01DataDt0Loaded != null, "expected load result to be non-null")
    }
  }
}

File: rust/tests/one_time_password_entity_test.rs
// Generated basic-flow test for the one_time_password entity (model-driven;
// mirrors the go TestEntity generator).

#![allow(unused_variables, unused_mut, unused_imports)]

mod common;

use std::rc::Rc;

use common::*;

use thesmsworks_sdk::core::helpers::{getp, getpath, ja, jo, now_ms, setp, to_map};
use thesmsworks_sdk::utility::voxgigstruct as vs;
use thesmsworks_sdk::{test_sdk, Entity, ThesmsworksEntity, ThesmsworksSDK, Value};

#[test]
fn one_time_password_entity_instance() {
    let testsdk = test_sdk(Value::Noval, Value::Noval);
    let ent = testsdk.one_time_password(Value::Noval);
    assert_eq!(ent.get_name(), "one_time_password");
}

#[test]
fn one_time_password_entity_stream() {
    // stream() runs the list op through the full pipeline and yields each
    // result item. Seed two entities via test mode; with the `streaming`
    // feature active it yields the feature's incremental items, else it
    // falls back to the materialised items — either way every item yields.
    let seed = jo(vec![(
        "entity",
        jo(vec![(
            "one_time_password",
            jo(vec![
                ("strm01", jo(vec![("id", Value::str("strm01"))])),
                ("strm02", jo(vec![("id", Value::str("strm02"))])),
            ]),
        )]),
    )]);

    let sdkopts = jo(vec![(
        "feature",
        jo(vec![("streaming", jo(vec![("active", Value::Bool(true))]))]),
    )]);

    let testsdk = test_sdk(seed.clone(), sdkopts);
    let ent = testsdk.one_time_password(Value::Noval);
    let items: Vec<Value> = ent
        .stream("list", Value::empty_map(), Value::empty_map())
        .expect("stream failed")
        .collect();
    assert_eq!(items.len(), 2, "stream should yield both seeded items");

    // Fallback: streaming inactive still yields both materialised items.
    let plainsdk = test_sdk(seed, Value::Noval);
    let plainent = plainsdk.one_time_password(Value::Noval);
    let plain_items: Vec<Value> = plainent
        .stream("list", Value::empty_map(), Value::empty_map())
        .expect("stream failed")
        .collect();
    assert_eq!(plain_items.len(), 2, "fallback stream should yield both items");
}

#[test]
fn one_time_password_entity_basic() {
    let setup = one_time_password_basic_setup(Value::Noval);
    // Per-op sdk-test-control.json skip — the basic test exercises a flow
    // with multiple ops; skipping any op skips the whole flow.
    let mode = if setup.live { "live" } else { "unit" };
    for op in ["create", "load"] {
        let (skip, reason) = is_control_skipped("entityOp", &format!("one_time_password.{}", op), mode);
        if skip {
            let reason = if reason.is_empty() {
                "skipped via sdk-test-control.json".to_string()
            } else {
                reason
            };
            eprintln!("skip: {}", reason);
            return;
        }
    }
    // The basic flow consumes synthetic IDs from the fixture. In live mode
    // without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only {
        eprintln!("skip: live entity test uses synthetic IDs from fixture — set THESMSWORKS_TEST_ONE_TIME_PASSWORD_ENTID JSON to run live");
        return;
    }
    let client = setup.client.clone();
    // CREATE
    let one_time_password_ref01_ent = client.one_time_password(Value::Noval);
    let one_time_password_ref01_data = to_map(&getp(
        &getpath(&["new", "one_time_password"], &setup.data),
        "one_time_password_ref01",
    ));

    let one_time_password_ref01_data_result = one_time_password_ref01_ent
        .create(one_time_password_ref01_data.clone(), Value::Noval)
        .expect("create failed");
    let one_time_password_ref01_data = to_map(&one_time_password_ref01_data_result);
    assert!(
        matches!(one_time_password_ref01_data, Value::Map(_)),
        "expected create result to be a map"
    );

    // LOAD
    let one_time_password_ref01_match_dt0 = Value::empty_map();
    let one_time_password_ref01_data_dt0_loaded = one_time_password_ref01_ent
        .load(one_time_password_ref01_match_dt0.clone(), Value::Noval)
        .expect("load failed");
    assert!(
        !one_time_password_ref01_data_dt0_loaded.is_noval(),
        "expected load result to be non-nil"
    );

}

fn one_time_password_basic_setup(extra: Value) -> EntityTestSetup {
    load_env_local();

    let mut entity_data_file = manifest_dir();
    entity_data_file.push("..");
    entity_data_file.push(".sdk");
    entity_data_file.push("test");
    entity_data_file.push("entity");
    entity_data_file.push("one_time_password");
    entity_data_file.push("OneTimePasswordTestData.json");

    let entity_data = read_json(&entity_data_file);

    let options = jo(vec![("entity", getp(&entity_data, "existing"))]);

    let client = test_sdk(options, extra.clone());

    // Generate idmap via transform, matching the TS pattern.
    let idmap = vs::transform(
        &ja(vec![Value::str("one_time_password01"), Value::str("one_time_password02"), Value::str("one_time_password03"), Value::str("otp01"), Value::str("otp02"), Value::str("otp03")]),
        &jo(vec![(
            "`$PACK`",
            ja(vec![
                Value::str(""),
                jo(vec![
                    ("`$KEY`", Value::str("`$COPY`")),
                    (
                        "`$VAL`",
                        ja(vec![
                            Value::str("`$FORMAT`"),
                            Value::str("upper"),
                            Value::str("`$COPY`"),
                        ]),
                    ),
                ]),
            ]),
        )]),
        None,
    )
    .unwrap_or_else(|_| Value::empty_map());

    // Detect ENTID env override before env_override consumes it. When live
    // mode is on without a real override, the basic test runs against
    // synthetic IDs from the fixture and 4xx's.
    let entid_env_raw = std::env::var("THESMSWORKS_TEST_ONE_TIME_PASSWORD_ENTID").unwrap_or_default();
    let idmap_overridden =
        !entid_env_raw.trim().is_empty() && entid_env_raw.trim().starts_with('{');

    let env = env_override(jo(vec![
        ("THESMSWORKS_TEST_ONE_TIME_PASSWORD_ENTID", idmap.clone()),
        ("THESMSWORKS_TEST_LIVE", Value::str("FALSE")),
        ("THESMSWORKS_TEST_EXPLAIN", Value::str("FALSE")),
        ("THESMSWORKS_APIKEY", Value::str("NONE")),
    ]));

    let idmap_resolved = match to_map(&getp(&env, "THESMSWORKS_TEST_ONE_TIME_PASSWORD_ENTID")) {
        Value::Map(m) => Value::Map(m),
        _ => to_map(&idmap),
    };

    let live = getp(&env, "THESMSWORKS_TEST_LIVE") == Value::str("TRUE");

    let client = if live {
        let merged = vs::merge(
            &ja(vec![jo(vec![("apikey", getp(&env, "THESMSWORKS_APIKEY"))]), extra]),
            None,
        );
        ThesmsworksSDK::new(to_map(&merged))
    } else {
        client
    };

    EntityTestSetup {
        client,
        data: entity_data,
        idmap: idmap_resolved,
        env: env.clone(),
        explain: getp(&env, "THESMSWORKS_TEST_EXPLAIN") == Value::str("TRUE"),
        live,
        synthetic_only: live && !idmap_overridden,
        now: now_ms(),
    }
}
